components/script/dom/gpu.rs

```diff
@@ -0,0 +1,168 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::compartments::InCompartment;
+use crate::dom::bindings::codegen::Bindings::GPUBinding::GPURequestAdapterOptions;
+use crate::dom::bindings::codegen::Bindings::GPUBinding::{self, GPUMethods, GPUPowerPreference};
+use crate::dom::bindings::error::Error;
+use crate::dom::bindings::refcounted::{Trusted, TrustedPromise};
+use crate::dom::bindings::reflector::{reflect_dom_object, DomObject, Reflector};
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::gpuadapter::GPUAdapter;
+use crate::dom::promise::Promise;
+use crate::task::TaskOnce;
+use dom_struct::dom_struct;
+use ipc_channel::ipc::{self, IpcSender};
+use ipc_channel::router::ROUTER;
+use js::jsapi::Heap;
+use std::rc::Rc;
+use webgpu::wgpu;
+use webgpu::{WebGPURequest, WebGPUResponse, WebGPUResponseResult, WebGPUThread};
+
+#[dom_struct]
+pub struct GPU {
+    reflector_: Reflector,
+}
+
+impl GPU {
+    pub fn new_inherited() -> GPU {
+        GPU {
+            reflector_: Reflector::new(),
+        }
+    }
+
+    pub fn new(global: &GlobalScope) -> DomRoot<GPU> {
+        reflect_dom_object(Box::new(GPU::new_inherited()), global, GPUBinding::Wrap)
+    }
+
+    fn get_wgpu_thread(&self) -> Option<WebGPUThread> {
+        self.global().as_window().webgpu_thread()
+    }
+}
+
+pub trait AsyncWGPUListener {
+    fn handle_response(&self, response: WebGPUResponse, promise: &Rc<Promise>);
+}
+
+struct WGPUResponse<T: AsyncWGPUListener + DomObject> {
+    trusted: TrustedPromise,
+    receiver: Trusted<T>,
+}
+
+impl<T: AsyncWGPUListener + DomObject> WGPUResponse<T> {
+    #[allow(unrooted_must_root)]
+    fn response(self, response: WebGPUResponseResult) {
+        // JSAutoRealm needs to be manually made.
+        // Otherwise, Servo will crash.
+        let promise = self.trusted.root();
+        match response {
+            Ok(response) => self.receiver.root().handle_response(response, &promise),
+            Err(error) => promise.reject_error(Error::Type(format!(
+                "Received error from WebGPU thread: {}",
+                error
+            ))),
+        }
+    }
+}
+
+pub fn response_async<T: AsyncWGPUListener + DomObject + 'static>(
```

Comment:
Our intention would be using `AsyncWGPUListener` trait for other WebGPU interfaces in the future, since there are a handful of promise calls in the spec.

Reply:
Ah ok. 

Would you have to update the state of those objects at some point based on what's happening in the WebGPU backend(what is currently named "WebGPUThread")?

If so, I think you're going to have to keep track of the various objects passed to the promise, either as `Dom<_>` or `WeakRef<_>`, and store those somewhere that is accessible from the object that is handling the incoming IPC message in the route. See for example how `MessagePort` are tracked with https://github.com/servo/servo/blob/2caa227e10369c4590cba5b6bceb85d907cd1f8c/components/script/dom/globalscope.rs#L205 and how there is only a single route used to manage various incoming messages, at https://github.com/servo/servo/blob/2caa227e10369c4590cba5b6bceb85d907cd1f8c/components/script/dom/globalscope.rs#L616

I guess for now it's fine to keep it as it is, however I'm not sure that you can scale the `response_async` pattern to handle all those interfaces. We'll see.